File: apps/web/types/audit.ts
export interface AuditLogEntry {
  id: string;
  action: string;
  actor_user_id?: string | null;
  resource_type: string;
  resource_id?: string | null;
  request_id?: string | null;
  source_ip?: string | null;
  user_agent?: string | null;
  old_value?: Record<string, unknown> | null;
  new_value?: Record<string, unknown> | null;
  timestamp: string;
}

File: apps/api/app/schemas/audit.py
"""Audit Log Schemas for Security Monitoring."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class AuditLogResponse(BaseModel):
    """Sanitized representation of an application audit trail record."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    actor_user_id: uuid.UUID | None = None
    actor_username: str | None = None
    action: str
    resource_type: str
    resource_id: str
    old_value: dict[str, Any] | None = None
    new_value: dict[str, Any] | None = None
    source_ip: str | None = None
    user_agent: str | None = None
    request_id: str | None = None
    timestamp: datetime


class AuditLogListResponse(BaseModel):
    """Paginated collection of security audit log records."""

    items: list[AuditLogResponse]
    total: int = Field(..., description="Total matching audit records")
    page: int = Field(..., description="Current page index (1-based)")
    limit: int = Field(..., description="Items per page")
    total_pages: int = Field(..., description="Total pages available")
